src/Tools/DependencyManagement/AbstractServiceProvider.php

```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ * AbstractServiceProvider class file.
+ *
+ * @package Automattic/WooCommerce/Tools/DependencyManagement
+ */
+
+namespace Automattic\WooCommerce\Tools\DependencyManagement;
```

Comment:
I wanted to have anything DI related under the same namespace, and I use `Tools` as a means to kind of say "all non-business logic stuff goes here". Anyway, if that's too much what about `Automattic\WooCommerce\DependencyManagement\ServiceProviders`?

Reply:
I'm just not sure there should be a distinction between things related to the dependency container and things related to the application itself. In a very real way these service providers _ARE_ the application (since they bootstrap it configure it) so it seems strange to relegate them to a sub-namespace. Does that make sense?

Reply:
Yep, that makes sense. I guess you agree with using `Automattic\WooCommerce\DependencyManagement` then?

Reply:
Haha, I'm still sold on `Automattic\WooCommerce\Providers` or `Automattic\WooCommerce\ServiceProviders` so maybe we leave that for others to give input on too!

Reply:
While asking further input is never a bad idea, let me elaborate on why I still think that we need a `DependencyManagement` subnamespace.

In the programming world the term "service provider" is used generically to refer to anything that... well, provides a service to client code. Usually this materializes as a class that you are allowed to instantiate at will and then you get methods like `getFoobarService`. Think about the infamous "service locator" pattern for example.

So if we drop a `ServiceProviders` namespace directly in our "root" namespace it can be confused with a place where it's allowed to put such generic service-providing classes. By using `DependencyManagement\ServiceProvider` instead, it's clear that it's something related to our dependency injection framework instead.

Another reason is that while now we have only the `Container` and the service provider classes, we might need to add additional DI related classes (and probably will need when we add support for messing with the DI in tests), and we'll need a place to put these - the `DependencyManagement` is the perfect place for that.

Reply:
I think the biggest hesitation for me is just that I'm not sure what classes other than service providers will be needed in the `DependencyManagement` domain. From my perspective we just need service providers and the container itself, and with the container in `Automattic\WooCommerce`, there doesn't seem to be anything else that might go into that namespace. What are you thinking might end up there? As far as test classes go, I wonder if those might be better off in `Automattic\WooCommerce\Tests\Framework`? Just food for thought!

Reply:
See `ExtendedContainer` for example, which is used as the underlying container for `Container` even outside tests. And while it seems that nothing else will be required for now, who knows, that could change in the future.